src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,237 @@
+                    // We don't have an existing incoming activity. Start a brand new one ourselves.
+
+                    ourActivity = RemotingActivitySource.StartActivity(ActivityInName, ActivityKind.Server, activityParentContext.ActivityContext);
+                }
+
+                SetStartingActivityAttributes(ourActivity, reqMsg);
+            }
+        }
+
+        public void ProcessMessageFinish(IMessage replyMsg, bool bCliSide, bool bAsync)
+        {
+            // This will be called:
+            // 1) On the server after we are done processing the call and are returning back to client
+            //    => Current activity should be the "ActivityIn", we need to stop it
+            //
+            // 2) On the client, when the server call returns
+            //    => Current activity should be the "ActivityOut", we need to stop it as well
+
+            try
+            {
+                if (this.options.Filter?.Invoke(replyMsg) == false)
+                {
+                    return;
+                }
+            }
+            catch (Exception e)
+            {
+                RemotingInstrumentationEventSource.Log.MessageFilterException(e);
+                return;
+            }
+
+            var act = Activity.Current;
+            if (act == null)
+            {
+                RemotingInstrumentationEventSource.Log.NullActivity(nameof(TelemetryDynamicSink), nameof(this.ProcessMessageFinish));
+                return;
+            }
+
+            bool validClientActivity = bCliSide && act.OperationName == ActivityOutName;
+            bool validServerActivity = !bCliSide && act.OperationName == ActivityInName;
+            if (validClientActivity || validServerActivity)
+            {
+                if (replyMsg is IMethodReturnMessage returnMsg)
+                {
+                    if (returnMsg.Exception == null)
+                    {
+                        act.SetStatus(Status.Ok);
+                    }
+                    else
+                    {
+                        act.SetStatus(Status.Unknown);
+
+                        var e = returnMsg.Exception;
+                        act.SetTag(AttributeExceptionType, e.GetType().FullName);
+                        act.SetTag(AttributeExceptionMessage, e.Message);
+                        act.SetTag(AttributeExceptionStackTrace, e.ToString());
+                    }
+                }
+
+                act.Stop();
+            }
+            else
+            {
+                RemotingInstrumentationEventSource.Log.InvalidCurrentActivity(
+                    nameof(TelemetryDynamicSink),
+                    nameof(this.ProcessMessageFinish),
+                    $"{ActivityInName} or {ActivityOutName}",
+                    act.OperationName);
+            }
+        }
+
+        private static void SetStartingActivityAttributes(Activity activity, IMessage msg)
+        {
+            if (activity != null && activity.IsAllDataRequested)
```

Comment:
It looks like for client case null check happens before coming into the method, so for client it is a double-check here. Maybe move the null check out of here and check before entering for the server call? That way you don't pay the method invocation for server for null, and client doesn't get hit twice when not null.

Reply:
Makes sense, fixed. Changed the signature to use `IMethodMessage` as well for clarity so that can use better API instead of looking things up in `Properties`.